### PR TITLE
Added module - Locaterisk

### PR DESCRIPTION
**Pull request by lavitsamson-tech**
Added connector for locaterisk api

## Summary by Sourcery

Add a LocateRisk automation module and scan report connector for polling CSV reports and pushing parsed events.

New Features:
- Introduce a LocateRisk module with configuration for API key, scan ID, and report URL.
- Add a LocateRisk scan report connector that polls the LocateRisk CSV export API, parses rows, and forwards events to intakes.

Enhancements:
- Align module and connector naming, manifest, and Docker packaging with repository conventions.
- Configure project metadata, dependencies, formatting, and test tooling for the LocateRisk module.

Tests:
- Add unit tests for the LocateRisk connector covering CSV parsing, empty rows, UTF-8 BOM handling, and HTTP error logging.
- Add pytest configuration and test data storage fixture for the LocateRisk module.

### CODE REVIEW

File: LocateRisk/locaterisk_modules/connector_locaterisk_scan_report.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            batch_of_events = []
+            try:
+                response = requests.get(
+                    f"{self.module.configuration.report_url}/{self.module.configuration.scan_id}/csv",
+                    headers={"Authorization": f"Bearer {self.module.configuration.api_key}"},
+                    timeout=60,
+                )
+
+                response.raise_for_status()
+                response.encoding = "utf-8-sig"  # handle UTF-8 BOM if present
```
[the remark this answers is not in the document]
Comment (lavitsamson-tech) [reply LocateRisk/locaterisk_modules/connector_locaterisk_scan_report.py:83]: resolved the issue

File: LocateRisk/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+[project.dependency-group.dev]
+pytest = "*"
+pytest-cov = "*"
+requests = "*"
+requests-mock = "*"
+
+[build-system]
+requires = ["uv_build>=0.9.7,<1.0.0"]
+build-backend = "uv_build"
+
```
[the remark this answers is not in the document]
Comment (lavitsamson-tech) [reply LocateRisk/pyproject.toml]: resolved issue

**Comment by mchupeau-sk** [comment]
Hello @lavitsamson-tech, a first review

## Blockers

- [manifest.json#L31 / pyproject.toml#L3] **Version mismatch**: `manifest.json` declares `"version": "0.1.9"` while `pyproject.toml` has `version = "0.1.8"`. These must agree. The latest dated CHANGELOG entry is `0.1.8`; either stamp the `## Unreleased` changes to `0.1.9` in the CHANGELOG (keeping `manifest.json`) and bump `pyproject.toml`, or roll the manifest back to `0.1.8`. → Choose one version, align all three files.

- [locaterisk_modules/models.py#L5] **`api_key` missing `Field(secret=True)`**: The `api_key` field is correctly listed in `manifest.json`'s `secrets` array, but the Pydantic model does not carry `secret=True`. Without this, the `sekoia-automation-sdk` won't mask the value in telemetry and log sink output. → Change to `api_key: str = Field(..., description="API Key", secret=True)`.

## Should-fix

- [locaterisk_modules/connector_locaterisk_scan_report.py#L71] **Unresponsive shutdown on long sleep**: The `run()` loop calls `time.sleep(self.configuration.polling_interval * 60)` (up to 5 minutes). When the SDK sets `self.running = False`, the process will not exit until the sleep completes. → Replace with `self._stop_event.wait(timeout=self.configuration.polling_interval * 60)`, which wakes immediately on stop.

- [locaterisk_modules/connector_locaterisk_scan_report.py#L33] **No retry / back-off on transient HTTP errors**: `requests.get` is called with no retry adapter. A single 429, 502, or 503 causes the entire polling cycle to be logged as an error and the data is lost until the next interval. → Mount a `urllib3.util.retry.Retry` adapter on a `requests.Session` (see conventions `client/` pattern), or at minimum catch `ConnectionError`/`Timeout` separately and back off.

- [pyproject.toml#L56] **`isort` missing from dev dependencies**: `[tool.isort]` is configured but `isort` is not listed under `[dependency-groups] dev`. CI or a local `uv run isort --check` will fail. → Add `"isort (>=5.13,<6)"` to the `dev` group.

## Nits

- [locaterisk_modules/] **No `metrics.py`**: Convention modules expose `prometheus_client` counters (`symphony_module_locaterisk_*`) for incoming events and forwarding duration. Not blocking for an initial connector, but worth adding for observability consistency.

## Notes

- **`LocateRisk/.gitignore`** is a straight copy of the GitHub Python `.gitignore` template (including Marimo, Streamlit, Abstra entries). Even if kept, it should be trimmed to only what is relevant to this module.

- **Dockerfile**: Uses `uv sync --frozen --no-install-project --no-dev --compile-bytecode` with a non-root `sekoiaio-runtime` user and `ENTRYPOINT ["python", "./main.py"]`. This follows the expected conventions for a uv-based module.

**Comment by lavitsamson-tech** [comment]
Fixed the issues.

**Comment by lavitsamson-tech** [comment]
docker file updated

**Comment by mchupeau-sk** [comment]
## 1. Blockers

### B1 — Connector descriptor is a raw Pydantic schema dump (missing `secrets`, `$schema`, and not normalized)

**Severity:** Blocker (security + correctness)  
**File:** `LocateRisk/connector_locaterisk_scan_report.json`

The `arguments` object is the raw output of Pydantic's `model_json_schema()` and was never normalized into a proper Sekoia connector descriptor. The tells are the auto-generated `"title"` keys, the schema title `"LocateRiskScanReportConnectorConfiguration"` (the literal Python class name), and the `anyOf: [string, null]` on `intake_server` (produced by the SDK's `intake_server: str | None = None`). Compared with `Okta/connector_pull_system_logs.json` and `Okta/connector_okta_user_assets.json`, four issues stand out:

1. **No `"secrets"` array** — `intake_key` is a sensitive credential and must be listed so the platform masks it in the UI and in stored configuration. (Okta lists `"secrets": ["sekoia_api_key"]`.)
2. **No `"$schema": "http://json-schema.org/draft-07/schema#"`** declaration.
3. **Pydantic artifacts instead of clean draft-07** — `intake_server` uses `anyOf` + `"default": null`. The default should be the real intake endpoint (`https://intake.sekoia.io`), not `null`, and the `"title"` keys should be replaced by `"description"` fields per repo convention.
4. **No trailing newline** (the diff shows `\ No newline at end of file`).

**Fix:** Replace the descriptor with a normalized draft-07 version modeled on the Okta descriptors:
```json
{
  "name": "LocateRisk Scan Report",
  "description": "Collect findings from a LocateRisk scan report export",
  "uuid": "d18a2265-324d-53d0-b2c8-e29748f1638f",
  "docker_parameters": "locaterisk_scan_report",
  "arguments": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "properties": {
      "intake_server": {
        "description": "Server of the intake server (e.g. 'https://intake.sekoia.io')",
        "default": "https://intake.sekoia.io",
        "type": "string"
      },
      "intake_key": {
        "description": "Intake key to use when sending events",
        "type": "string"
      },
      "polling_interval": {
        "description": "Polling interval in minutes",
        "default": 5,
        "type": "integer"
      }
    },
    "required": ["intake_key"],
    "secrets": ["intake_key"],
    "title": "LocateRisk Scan Report configuration",
    "type": "object"
  }
}
```
(ensure the file ends with a newline)

---

### B2 — `report_url` not validated to use HTTPS

**Severity:** Blocker (security)  
**File:** `LocateRisk/locaterisk_modules/models.py`, line 6–9

`report_url` is a plain `str` with a default `https://` URL, but nothing prevents a user (or operator) from setting it to an `http://` URL. The `Authorization: Bearer <api_key>` header would then be sent in cleartext.

Additionally, the retry adapter in the connector is also mounted on `http://`:
```python
# LocateRisk/locaterisk_modules/connector_locaterisk_scan_report.py, lines 40–41
self._session.mount("https://", adapter)
self._session.mount("http://", adapter)   # ← unnecessary and dangerous
```

**Fix :** Use Pydantic's `AnyHttpsUrl` type to reject non-HTTPS URLs at configuration time:
```python
from pydantic import AnyHttpsUrl, Field

class LocateRiskModuleConfiguration(BaseModel):
    ...
    report_url: AnyHttpsUrl = Field(
        "https://app.locaterisk.com/api/rest/report/export",
        description="Base URL of the LocateRisk report export endpoint",
    )
```

---


### B3 — No deduplication / checkpoint

**Severity:** Should-fix  
**File:** `LocateRisk/locaterisk_modules/connector_locaterisk_scan_report.py`, `run()` method

The connector fetches the full CSV on every poll and pushes **all rows** each time. If a report does not change between polls, every row will be pushed again. There is no cursor, row hash, or etag-based conditional request.

**Fix:** Use an SDK checkpoint instead of an ad-hoc approach. Reasonable options for a CSV-export source:
- Persist a content hash of the last report (`hashlib.sha256(response.text.encode()).hexdigest()`) in a `PersistentJSON("context.json", self._data_path)` and skip the push when unchanged; update it **after** a successful push.
- If the export carries a timestamp column, use `CheckpointDatetime`/`CheckpointTimestamp` (with `start_at` + `ignore_older_than`) keyed on the latest row seen, optionally with a bounded `LRUCache` of row IDs persisted in the same context to dedup rows sharing a timestamp.

See the patterns in `1Password/onepassword_modules/connector_1password_epm.py` (`CheckpointDatetime` + `subkey`), `Akamai/akamai_modules/connector_akamai_waf.py` (`CheckpointTimestamp` + in-context dedup cache), and `AWS/asset_connector/aws_assets.py` (`PersistentJSON` `context.json`).

---

### B4 — `sekoia-automation-sdk` has no version constraint

**Severity:** Should-fix  
**File:** `LocateRisk/pyproject.toml`, line 9

```toml
"sekoia-automation-sdk",   # ← no version pin
```

All other dependencies have constraints. An unpinned SDK means any breaking SDK update could silently break this module.

**Fix:** Add a lower bound at minimum:
```toml
"sekoia-automation-sdk>=X.Y,<Z"
```
(replace `X.Y` with the current known-good version; check `uv.lock` for the locked version).

---

### B5 — `connector_locaterisk_scan_report.json` missing trailing newline

**File:** `LocateRisk/connector_locaterisk_scan_report.json`, last line  
The diff shows `\ No newline at end of file`. All text files in the repo should end with a newline.

---

### B6 — Dockerfile copies test artifacts into the image

**File:** `LocateRisk/Dockerfile`  
`COPY . .` copies `tests/`, `.gitignore`, `CHANGELOG.md`, etc. into the production image. Add a `.dockerignore` file:
```
tests/
.gitignore
CHANGELOG.md
*.md
.mypy_cache/
.pytest_cache/
htmlcov/
.coverage
```

---

### B7 — `manifest.json` `categories` is an empty array

**File:** `LocateRisk/manifest.json`, line 29  
`"categories": []` — assign the appropriate category (e.g., `["Threat & Vulnerability Intelligence"]`) so the module is discoverable in the platform.

---

**Comment by mchupeau-sk** [comment]
@lavitsamson-tech can you check this [comment ](https://github.com/SEKOIA-IO/automation-library/pull/2677#issuecomment-4780121886)

**Comment by lavitsamson-tech** [comment]
1. Connector descriptor normalized — LocateRisk/connector_locaterisk_scan_report.json
Rewrote the raw Pydantic model_json_schema() dump into a proper draft-07 descriptor: added $schema, replaced auto-generated title keys with description fields, dropped the anyOf: [string, null] artifact on intake_server and gave it a real default (https://intake.sekoia.io).
Added "secrets": ["intake_key"] so the platform masks the intake key in the UI and stored config.
Added the missing trailing newline.

2. HTTPS enforcement (security) — LocateRisk/locaterisk_modules/models.py and connector_locaterisk_scan_report.py
report_url now rejects non-HTTPS URLs at configuration time via a @field_validator.
Removed the http:// retry-adapter mount from the connector's session, so the API-key Bearer token can never go out in cleartext.

3. Row-level deduplication / checkpoint — connector_locaterisk_scan_report.py
The connector persists SHA-256 hashes of forwarded rows in a context.json checkpoint (PersistentJSON) and only pushes rows not seen on the previous poll; an unchanged report is no longer re-sent every cycle.
Reworded the "no events" log to "No new events to push this cycle".

4. Tests for the new behavior — LocateRisk/tests/test_connector.py
Added test_run_deduplicates_unchanged_rows_across_polls and test_run_forwards_only_new_rows.

5. SDK dependency pinned — LocateRisk/pyproject.toml (+ uv.lock re-synced)
sekoia-automation-sdk constrained to >=1.23.0,<2.0.0

6. Docker image slimmed — LocateRisk/.dockerignore (new file)
Excludes tests, coverage artifacts, caches, docs, and the local .venv/ (which could otherwise clobber the image's /app/.venv) from the build context.

7. Module discoverability — LocateRisk/manifest.json
Set categories to ["Threat Intelligence"] (was empty).

**Review by mchupeau-sk — APPROVED** [review]
LGFM
